File: day11/main2.py
import sys
import os
sys.setrecursionlimit(10**7)

def make_step(num):
    # Rule 1: If stone is 0 -> becomes stone with number 1
    if num == 0:
        return 1, None
    # Rule 2: If the stone number has an even number of digits -> split
    elif len(str(num)) % 2 == 0:
        str_num = str(num)
        half = len(str_num) // 2
        left_part = int(str_num[:half])
        right_part = int(str_num[half:])
        return left_part, right_part
    # Rule 3: Otherwise, multiply by 2024
    else:
        return num * 2024, None

def step_through(stones, blinks):
    for i in range(blinks):
        new_stones = []
        for num in stones:
            step, new_num = make_step(num)
            new_stones.append(step)
            if new_num is not None:
                new_stones.append(new_num)
        stones = new_stones
        print(f"blink {i}, {len(stones)} stones")
    return stones


def main():
    directory = os.path.dirname(__file__) + '\\'
    # Example input
    input = [64554, 35, 906, 6, 6960985, 5755, 975820, 0]
    input2 = [0, 1, 10, 99, 999]
    input3 = [125, 17]

    input = input
    blinks = 75

    for number in input:
        result = step_through([number], blinks)
        print(result)
        print(len(result))
        with open(directory + 'output.txt', 'w') as f:
            f.write(str(len(result)))

if __name__ == "__main__":
    main()

File: day11/main2.go
package main

import (
	"fmt"
	"strconv"
)

func makeStep(num int) (int, *int) {
	s := strconv.Itoa(num)
	// Rule 1
	if num == 0 {
		return 1, nil
	}
	// Rule 2
	if len(s)%2 == 0 {
		half := len(s) / 2
		leftPart, _ := strconv.Atoi(s[:half])
		rightPart, _ := strconv.Atoi(s[half:])
		return leftPart, &rightPart
	}
	// Rule 3
	res := num * 2024
	return res, nil
}

func stepThrough(stones []int, blinks int) []int {
	for i := 0; i < blinks; i++ {
		newStones := make([]int, 0, len(stones)*2) // Guessing a bigger capacity
		for _, num := range stones {
			step, newNum := makeStep(num)
			newStones = append(newStones, step)
			if newNum != nil {
				newStones = append(newStones, *newNum)
			}
		}
		stones = newStones
		// Optionally print progress
		fmt.Printf("After %d blinks: %d stones\n", i+1, len(stones))
	}
	return stones
}

func main() {
	// Example initial configuration - adjust as needed.
	// For a real puzzle input, read from file or standard input.
	//stones := []int{0, 1, 10, 99, 999} // just an example
	stones := []int{64554, 35, 906, 6, 6960985, 5755, 975820, 0} // just an example
	blinks := 75

	// Run the process
	finalStones := stepThrough(stones, blinks)

	fmt.Println("Final number of stones:", len(finalStones))
}
